dsls/src/main/scala/ch/epfl/directembedding/test/TestBase.scala

```diff
@@ -17,11 +17,35 @@ case class Take[T](self: Exp[_], n: Exp[Int]) extends Exp[TArgClassExample[T]]
 case class X[T](self: Exp[_]) extends Exp[Int]
 case class Y[T](self: Exp[_]) extends Exp[Int]
 case class TArgsZ[T, U](self: Exp[_]) extends Exp[U]
-case class AppCurry[T](self: Exp[_], p1: Exp[T], p2: Exp[T], p3: Exp[T]*) extends Exp[T]
-case class AppManyArgs[T](self: Exp[_], p1: Exp[T]*) extends Exp[T]
+case class AppCurry[T](self: Exp[_], p1: Exp[_], p2: Exp[_], p3: Exp[_]*) extends Exp[T]
+case class AppManyArgs[T](self: Exp[_], p1: Exp[_]*) extends Exp[T]
 
-case class IF[T](cond: Exp[Boolean], e1: Exp[T], e2: Exp[T]) extends Exp[T]
+case class IF[T](cond: Exp[Boolean], e1: Exp[_], e2: Exp[_]) extends Exp[T]
+case class While[T](cond: Exp[Boolean], `do`: Exp[_]) extends Exp[T]
+case class DoWhile[T](cond: Exp[Boolean], `do`: Exp[_]) extends Exp[T]
+case class Return[T](cond: Exp[T]) extends Exp[T]
 case class NewVar[T](e: Exp[T]) extends Exp[T]
+case class ReadVar[T](e: Exp[T]) extends Exp[T]
+case class Assign[T](x: Exp[T], t: Exp[_]) extends Exp[T]
+case class Infix_==[T](t: Exp[_], t1: Exp[_]) extends Exp[T]
+case class Infix_Equals[T](t: Exp[T], t1: Exp[_]) extends Exp[T]
+case class Infix_Eq[T](t: Exp[_], t1: Exp[_]) extends Exp[T]
+case class Infix_Ne[T](t: Exp[_], t1: Exp[_]) extends Exp[T]
+case class Infix_!=[T](t: Exp[_], t1: Exp[_]) extends Exp[T]
+case class Infix_##[T](t: Exp[T]) extends Exp[T]
+case class Infix_HashCode[T](t: Exp[T]) extends Exp[T]
+case class Infix_IsInstanceOf[T](t: Exp[_]) extends Exp[T]
```

Comment:
The same issue as with `synchronized`. In my opinion, the transformation should provide the type of `t` as well as the type argument to `isInstanceOf[T]`. The signature of this class would then be something like this:

```scala
case class Infix_IsInstanceOf[T, U](t: Exp[T]) extends Exp[Boolean]
```
@vjovanov